internal/codeintel/dependencies/service.go

```diff
@@ -13,118 +12,174 @@ import (
+		lockfilesSemaphore: semaphore.NewWeighted(int64(lockfilesSemaphoreWeight)),
+		syncer:             syncer,
+		syncerSemaphore:    semaphore.NewWeighted(int64(syncerSemaphoreWeight)),
+		operations:         newOperations(observationContext),
 	}
 }
 
 // Dependencies resolves the (transitive) dependencies for a set of repository and revisions.
 // Both the input repoRevs and the output dependencyRevs are a map from repository names to revspecs.
 func (s *Service) Dependencies(ctx context.Context, repoRevs map[api.RepoName]types.RevSpecSet) (dependencyRevs map[api.RepoName]types.RevSpecSet, err error) {
-	logFields := make([]log.Field, 0, 2)
-	if len(repoRevs) == 1 {
-		for repoName, revs := range repoRevs {
-			revStrs := make([]string, 0, len(revs))
-			for rev := range revs {
-				revStrs = append(revStrs, string(rev))
-			}
-
-			logFields = append(logFields,
-				log.String("repo", string(repoName)),
-				log.String("revs", strings.Join(revStrs, ",")),
-			)
-		}
-	} else {
-		logFields = append(logFields, log.Int("repoRevs", len(repoRevs)))
-	}
-
-	ctx, endObservation := s.operations.dependencies.With(ctx, &err, observation.Args{LogFields: logFields})
+	ctx, endObservation := s.operations.dependencies.With(ctx, &err, observation.Args{LogFields: constructLogFields(repoRevs)})
 	defer func() {
 		endObservation(1, observation.Args{LogFields: []log.Field{
 			log.Int("numDependencyRevs", len(dependencyRevs)),
 		}})
 	}()
 
-	var mu sync.Mutex
-	dependencyRevs = make(map[api.RepoName]types.RevSpecSet)
+	// Parse lockfile contents for the given repository and revision pairs
+	deps, err := s.lockfileDependencies(ctx, repoRevs)
+	if err != nil {
+		return nil, err
+	}
 
-	sem := semaphore.NewWeighted(32)
-	g, ctx := errgroup.WithContext(ctx)
+	// Populate return value map from the given information. In teh same pass, populate
+	// auxiliary data structures that can be used to feed the upsert and sync operations
+	// below.
+	dependencyRevs = make(map[api.RepoName]types.RevSpecSet, len(repoRevs))
+	dependencies := []store.DependencyRepo{}
+	repoNamesByDependency := map[store.DependencyRepo]api.RepoName{}
+
+	for _, dep := range deps {
+		repo := dep.RepoName()
+		rev := api.RevSpec(dep.GitTagFromVersion())
+		scheme := dep.Scheme()
+		name := dep.PackageSyntax()
+		version := dep.PackageVersion()
+
+		if _, ok := dependencyRevs[repo]; !ok {
+			dependencyRevs[repo] = types.RevSpecSet{}
+		}
+		dependencyRevs[repo][rev] = struct{}{}
+
+		dependencyRepo := store.DependencyRepo{Scheme: scheme, Name: name, Version: version}
+		dependencies = append(dependencies, dependencyRepo)
+		repoNamesByDependency[dependencyRepo] = repo
+	}
 
+	// Write depenencies to database and sync all of the ones that were newly inserted
+	newDependencies, err := s.dependenciesStore.UpsertDependencyRepos(ctx, dependencies)
+	if err != nil {
+		return nil, err
+	}
+	if err := s.syncNew(ctx, newDependencies, repoNamesByDependency); err != nil {
```

Comment:
There's nothing about the method that is coupled with a dependency being new.

```suggestion
	if err := s.sync(ctx, newDependencies, repoNamesByDependency); err != nil {
```

Reply:
If we passed a slice of `[]reposource.PackageDependency` here, we wouldn't need the additional map, since each dependency knows its own repo name.